File: experiments/016-LanguageComparisonRepeated/cells/java-gradle_1-Empty_free/2024-1c-parcial1/20260917-062519-53824/analysis/acceptance-project/src/main/java/simshotels/Hotel.java
package simshotels;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.ToIntFunction;

public class Hotel {

    private final ArrayList<Floor> floors;

    // instance creation (class side)

    public static Hotel withFloors(List<Floor> aFloorsCollection) {
        Hotel.assertHaveFloors(aFloorsCollection);

        return new Hotel(aFloorsCollection);
    }

    // initialization

    private Hotel(List<Floor> aFloorsCollection) {
        floors = new ArrayList<>(aFloorsCollection);
    }

    // accounting

    public int totalLosses() {
        return sumOverFloors(Floor::totalLosses);
    }

    public int totalProfits() {
        return sumOverFloors(Floor::totalProfits);
    }

    // totals

    public int totalRooms() {
        return sumOverFloors(Floor::totalRooms);
    }

    public int totalRoomsAvailable() {
        return sumOverFloors(Floor::totalRoomsAvailable);
    }

    public int totalRoomsOccupied() {
        return sumOverFloors(Floor::totalRoomsOccupied);
    }

    public int totalRoomsReserved() {
        return sumOverFloors(Floor::totalRoomsReserved);
    }

    private int sumOverFloors(ToIntFunction<Floor> aFloorValueBlock) {
        return floors.stream().mapToInt(aFloorValueBlock).sum();
    }

    // testing

    public boolean isEmpty() {
        return floors.stream().allMatch(Floor::isAvailable);
    }

    // guests

    public void receiveAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        withFloorNumberedDo(aFloorNumber, floor -> floor.receiveAtRoom(aGuestType, aRoomNumber));
    }

    public void receiveWithReservationAtFloorAtRoom(String aGuestType, int aFloorNumber, int aRoomNumber) {
        withFloorNumberedDo(aFloorNumber, floor -> floor.receiveWithReservationAtRoom(aGuestType, aRoomNumber));
    }

    public void reserveRoomAtFloor(int aRoomNumber, int aFloorNumber) {
        withFloorNumberedDo(aFloorNumber, floor -> floor.reserveRoom(aRoomNumber));
    }

    private void withFloorNumberedDo(int aFloorNumber, Consumer<Floor> aFloorBlock) {
        if (aFloorNumber < 1 || aFloorNumber > floors.size()) {
            Hotel.signalFloorNumberDoesNotExist();
        }
        aFloorBlock.accept(floors.get(aFloorNumber - 1));
    }

    // assertions (class side)

    public static void assertHaveFloors(List<Floor> aFloorsCollection) {
        if (aFloorsCollection.isEmpty()) {
            Hotel.signalNoFloors();
        }
    }

    // error description (class side)

    public static String floorNumberDoesNotExistErrorDescription() {
        return "Floor number does not exist";
    }

    public static String noFloorsErrorDescription() {
        return "Cannot have a Hotel without floors";
    }

    // exceptions (class side)

    public static void signalFloorNumberDoesNotExist() {
        throw new RuntimeException(Hotel.floorNumberDoesNotExistErrorDescription());
    }

    public static void signalNoFloors() {
        throw new RuntimeException(Hotel.noFloorsErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2024-1c-parcial1/20260921-205145-86356/analysis/own-tests-project/src/simshotels/hotel.py
from simshotels.floor import Floor


class Hotel:

    # instance creation (class side)

    @classmethod
    def with_floors(cls, a_floors_collection):
        cls.assert_have_floors(a_floors_collection)

        return cls(a_floors_collection)

    # initialization

    def __init__(self, a_floors_collection):
        self._floors = a_floors_collection

    # testing

    def is_empty(self):
        return all(floor.is_available() for floor in self._floors)

    # guests

    def receive_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_at_room(a_guest_type, a_room_number)

    def receive_with_reservation_at_floor_at_room(self, a_guest_type, a_floor_number, a_room_number):
        self.floor_at(a_floor_number).receive_with_reservation_at_room(a_guest_type, a_room_number)

    def reserve_room_at_floor(self, a_room_number, a_floor_number):
        self.floor_at(a_floor_number).reserve_room(a_room_number)

    # private

    def floor_at(self, a_floor_number):
        if a_floor_number < 1 or a_floor_number > self.total_floors():
            Hotel.signal_floor_number_does_not_exist()

        return self._floors[a_floor_number - 1]

    # accounting

    def total_losses(self):
        return self.total_of(lambda floor: floor.total_losses())

    def total_profits(self):
        return self.total_of(lambda floor: floor.total_profits())

    # totals

    def total_floors(self):
        return len(self._floors)

    def total_rooms(self):
        return self.total_of(lambda floor: floor.total_rooms())

    def total_rooms_available(self):
        return self.total_of(lambda floor: floor.total_rooms_available())

    def total_rooms_occupied(self):
        return self.total_of(lambda floor: floor.total_rooms_occupied())

    def total_rooms_reserved(self):
        return self.total_of(lambda floor: floor.total_rooms_reserved())

    def total_of(self, a_floor_amount):
        return sum(map(a_floor_amount, self._floors))

    # assertions (class side)

    @staticmethod
    def assert_have_floors(a_floors_collection):
        if len(a_floors_collection) == 0:
            Hotel.signal_no_floors()

    # error description (class side)

    @staticmethod
    def floor_number_does_not_exist_error_description():
        return "Floor number does not exist"

    @staticmethod
    def no_floors_error_description():
        return "Cannot have a Hotel without floors"

    @staticmethod
    def room_number_does_not_exist_error_description():
        # the room numbers of a floor are known by the floor
        return Floor.room_number_does_not_exist_error_description()

    # exceptions (class side)

    @staticmethod
    def signal_floor_number_does_not_exist():
        raise RuntimeError(Hotel.floor_number_does_not_exist_error_description())

    @staticmethod
    def signal_no_floors():
        raise RuntimeError(Hotel.no_floors_error_description())
